cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -221,14 +225,42 @@ protected void startUp() throws Exception {
         SparkRuntimeEnv.setProperty(key, sparkDefaultConf.getProperty(key));
       }
 
-      if (isLocal) {
+      if (masterEnv != null) {
+        // Only add cconf, hconf for master environment
+        localizeResources.add(new LocalizeResource(saveCConf(cConfCopy, tempDir)));
+        Configuration hConf = contextConfig.set(runtimeContext, pluginArchive).getConfiguration();
+        localizeResources.add(new LocalizeResource(saveHConf(hConf, tempDir)));
+
+        // Localize all the files from user resources
+        List<File> files = copyUserResources(context.getLocalizeResources(), tempDir);
+        for (File file : files) {
+          localizeResources.add(new LocalizeResource(file));
+        }
+        // Master environment would always contain this file.
+        // https://github.com/cdapio/cdap/blob/develop/cdap-spark-core3_2.12/src/k8s/Dockerfile#L46
+        jobFile = URI.create("local:/opt/cdap/cdap-spark-core/cdap-spark-core.jar");
```

Comment:
should we move it to master env spark submitter?

Reply:
yeah, shouldn't be here as it's specific to k8s and isn't something that needs to be localized. One way would be to add a field to the SparkConfig class and add some method to the AbstractSparkSubmitter that MasterEnvSparkSubmitter can override.

Reply:
sounds good 👍🏼